src/main/scala/org/broadinstitute/dsde/firecloud/service/TrialService.scala

```diff
@@ -71,110 +72,128 @@ final class TrialService
-                                userInfo: WorkbenchUserInfo,
-                                statusTransition: UserTrialStatus => UserTrialStatus): Future[StatusUpdate.Attempt] = {
-    // get the status of the "userInfo" user, but authenticating to Thurloe as the "managerInfo" user
-    thurloeDao.getTrialStatus(userInfo.userSubjectId, managerInfo) flatMap {
-      case Some(currentStatus) => {
-        val currentState = currentStatus.state
-        val newStatus = statusTransition(currentStatus)
-        val newState = newStatus.state
-
-        if (currentState == newState) {
-          logger.warn(
-            s"The user '${userInfo.userEmail}' is already in the trial state of '$newState'. " +
-              s"No further action will be taken.")
-
-          Future(StatusUpdate.Success)
-        } else {
-          logger.warn(s"Current trial state of the user '${userInfo.userEmail}' is '$currentState'")
-          logger.warn(s"Checking if we can transition from $currentState to $newState...")
-
-          require(newState.nonEmpty, "Cannot transition to an unspecified state")
-
-          // TODO: Handle invalid initial trial status by
-          //    -adding another case to Trial.StatusUpdate
-          //    -using that case to return a BadRequest from the parent method (i.e. updateUserState())
-          assert(newState.get.isAllowedFrom(currentState),
-            s"Cannot transition from $currentState to $newState")
-
-          logger.warn(s"Yes, we can! Proceeding...")
-
-          // TODO: Test the logic below by adding a mock ThurloeDAO whose saveTrialStatus() returns Failure
-          // Save updates to "userInfo's" trial status, authenticating to Thurloe as "managerInfo" user
-          thurloeDao.saveTrialStatus(userInfo.userSubjectId, managerInfo, newStatus) map {
-            case Success(_) =>
-              logger.warn(s"Updated profile saved as $newState; we are done!")
-              StatusUpdate.Success
-            case Failure(ex) =>
-              logger.warn(s"We could not have Thurloe update the new state as $newState; aborting...")
-              StatusUpdate.ServerError(ex.getMessage)
+  private def executeStateTransitions(managerInfo: UserInfo, userEmails: Seq[String],
+                                      statusTransition: (WorkbenchUserInfo, Option[UserTrialStatus]) => UserTrialStatus,
+                                      transitionPostProcessing: (Attempt, Option[UserTrialStatus], UserTrialStatus) => Unit): Future[PerRequestMessage] = {
+    var results: Seq[(String,String)] = Seq()
+    userEmails.foreach { userEmail =>
+     val finalStatus =
+      getUserSubjectId(userEmail) flatMap { subId =>
+        val userInfo = WorkbenchUserInfo(subId, userEmail)
+        thurloeDao.getTrialStatus(subId, managerInfo) flatMap { userTrialStatus =>
+          val newStatus = statusTransition(userInfo, userTrialStatus)
+          if (requiresStateTransition(userTrialStatus, newStatus)) {
+            for {
```

Comment:
Nit: mixing a single-line for-comprehension seems out of place in the nested flatmaps.
```
            updateTrialStatus(managerInfo, userInfo, newStatus).map { stateResponse =>
              transitionPostProcessing(stateResponse, userTrialStatus, newStatus)
              StatusUpdate.toName(stateResponse)
            }
```